modules/search/class.jetpack-search.php

```diff
@@ -0,0 +1,1289 @@
+<?php
+
+class Jetpack_Search {
+
+	protected $found_posts = 0;
+
+	/**
+	 * The maximum offset ('from' param), since deep pages get exponentially slower.
+	 *
+	 * @see https://www.elastic.co/guide/en/elasticsearch/guide/current/pagination.html
+	 */
+	protected $max_offset = 200;
+
+	protected $search_result;
+
+	protected $original_blog_id;
+	protected $jetpack_blog_id;
+
+	protected $aggregations = array();
+	protected $max_aggregations_count = 100;
+
+	protected static $instance;
+
+	//Languages with custom analyzers, other languages are supported,
+	// but are analyzed with the default analyzer.
+	public static $analyzed_langs = array( 'ar', 'bg', 'ca', 'cs', 'da', 'de', 'el', 'en', 'es', 'eu', 'fa', 'fi', 'fr', 'he', 'hi', 'hu', 'hy', 'id', 'it', 'ja', 'ko', 'nl', 'no', 'pt', 'ro', 'ru', 'sv', 'tr', 'zh' );
+
+	protected function __construct() {
+		/* Don't do anything, needs to be initialized via instance() method */
+	}
+
+	public function __clone() {
+		wp_die( "Please don't __clone Jetpack_Search" );
+	}
+
+	public function __wakeup() {
+		wp_die( "Please don't __wakeup Jetpack_Search" );
+	}
+
+	/**
+	 * Get singleton instance of Jetpack_Search
+	 *
+	 * Instantiates and sets up a new instance if needed, or returns the singleton
+	 *
+	 * @module search
+	 *
+	 * @return Jetpack_Search The Jetpack_Search singleton
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new Jetpack_Search();
+
+			self::$instance->setup();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Perform various setup tasks for the class
+	 *
+	 * Checks various pre-requisites and adds hooks
+	 *
+	 * @module search
+	 */
+	public function setup() {
```

Comment:
Should there be a check here to ensure the site is at the right plan level or confirm that an ES index exists? I assume the ES check is too expensive to run as often as this would run, but add an option to indicate index is pending or set that is checked on a regular basis?